concourse/steps/update_component_deps.py

```
@@ -377,6 +382,163 @@ def _import_release_notes(
     return release_notes
 
 
+def format_bom_diff(
+    component_diff: cnudie.util.ComponentDiff,
+    to_component_name: str,
+    base_url: str | None
+):
+    class ChangeType(enum.StrEnum):
+        ADDED = 'added'
+        REMOVED = 'removed'
+        CHANGED = 'changed'
+
+    class ComponentChange:
+        def __init__(self, name, old_version=None, new_version=None, change_type=None):
+            self.name = name
+            self.old_version = old_version
+            self.new_version = new_version
+            self.change_type = change_type  # 'added', 'removed', 'changed'
+
+        def __str__(self):
+            if self.change_type == ChangeType.CHANGED:
+                return f'\U00002699 {self.name}: {self.old_version} → {self.new_version}'
+            elif self.change_type == ChangeType.REMOVED:
+                return f'\U00002796 {self.name} (v{self.old_version})'
+            elif self.change_type == ChangeType.ADDED:
+                return f'\U00002795 {self.name} (v{self.new_version})'
+            return ''
```

raise ValueError upon unexpected value rather than silently returning an empty string